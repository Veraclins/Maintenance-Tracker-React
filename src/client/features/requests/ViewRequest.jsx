import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import toastr from 'toastr';
import { PropTypes } from 'prop-types';
import history from '../../shared/utilities/history';
import { getSingleRequest } from './user-requests/userRequestsAction';


export class ViewRequest extends Component {
  componentDidMount() {
    const {
      match, location, fetchRequest, isLoggedIn, user,
    } = this.props;
    if (!isLoggedIn) {
      toastr.error('You must be logged to rate an article');
      return history.push('/login', { from: location.pathname });
    }
    const { params } = match;
    return fetchRequest(user, params.requestId);
  }

  render() {
    const { request } = this.props;
    return (
      <div className="request_content_layout body_gradient">
        <div className="container">
          {request.id
            ? (
              <div className="single_request_column">
                <div className="single_request_info">
                  <time>{moment(request.created_at).format('Do_MMMM_YYYY')}</time>
                  <span className="request_status">{request.status}</span>
                  <h3 className="single_request_title">{request.title}</h3>
                  <h4>{request.device}</h4>
                  <p className="single_request_body">
                    {request.description}
                  </p>
                  <Link to={`/requests/${request.id}/update`} className="edit_request_link"> <span>EDIT</span></Link>
                </div>
              </div>
            ) : <h3 className="single_request_title">No request with the given id</h3>}
        </div>
      </div>
    );
  }
}


ViewRequest.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  user: PropTypes.shape({}).isRequired,
  fetchRequest: PropTypes.func.isRequired,
  location: PropTypes.shape({}).isRequired,
  match: PropTypes.shape({}).isRequired,
  request: PropTypes.shape({}).isRequired,
};

const mapStateToProps = state => ({
  isLoggedIn: state.auth.isAuthenticated,
  user: state.auth.user,
  location: state.router.location,
  request: state.requests.currentRequest,
});
const mapDispatchToProps = dispatch => ({
  fetchRequest: (user, requestId) => dispatch(getSingleRequest(user, requestId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ViewRequest);
