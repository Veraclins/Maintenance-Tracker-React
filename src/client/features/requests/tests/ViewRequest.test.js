import React from 'react';
import { shallow } from 'enzyme';
import { Link } from 'react-router-dom';

import { ViewRequest } from '../ViewRequest';

const request = {
  id: 13,
  user_id: 2,
  title: 'General repainting',
  device: 'Smartphone',
  description: 'Although a downpour briefly interrupted play in the early stages of the second half, Nigeria played with more intent after half-time and went close in the 63rd minute, as John Obi Mikel saw a header saved by Vaclik.',
  status: 'pending',
  created_at: '2018-10-06T23:14:25.294Z',
  updated_at: '2018-10-06T23:14:25.294Z',
};

const mockFunction = jest.fn();

describe('Test the ViewRequest component', () => {
  it('renders with the right view for a user', () => {
    const isAdmin = false;
    const wrapper = shallow(<ViewRequest
      request={request}
      isAdmin={isAdmin}
      updateRequest={mockFunction}
    />);
    expect(wrapper.find('div')).toHaveLength(3);
    expect(wrapper.find(Link)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders with the right view for an admin', () => {
    const isAdmin = true;
    const wrapper = shallow(<ViewRequest
      request={request}
      isAdmin={isAdmin}
      updateRequest={mockFunction}
    />);
    expect(wrapper.find('div')).toHaveLength(3);
    expect(wrapper.find('a')).toHaveLength(2);
    expect(wrapper).toMatchSnapshot();
  });
});
