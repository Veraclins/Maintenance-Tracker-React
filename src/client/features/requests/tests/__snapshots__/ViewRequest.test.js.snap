// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the ViewRequest component renders with the right view for a user 1`] = `
<div
  className="single_request_column"
>
  <div
    className="single_request_info"
  >
    <time>
      7th_October_2018
    </time>
    <span
      className="request_status"
    >
      pending
    </span>
    <h3
      className="single_request_title"
    >
      General repainting
    </h3>
    <h4>
      Smartphone
    </h4>
    <p
      className="single_request_body"
    >
      Although a downpour briefly interrupted play in the early stages of the second half, Nigeria played with more intent after half-time and went close in the 63rd minute, as John Obi Mikel saw a header saved by Vaclik.
    </p>
    <div
      className="request_footer"
    >
      <Link
        className="edit_request_link"
        replace={false}
        to="/requests/13/edit"
      >
        <span>
          EDIT
        </span>
      </Link>
    </div>
  </div>
</div>
`;

exports[`Test the ViewRequest component renders with the right view for an admin 1`] = `
<div
  className="single_request_column"
>
  <div
    className="single_request_info"
  >
    <time>
      7th_October_2018
    </time>
    <span
      className="request_status"
    >
      pending
    </span>
    <h3
      className="single_request_title"
    >
      General repainting
    </h3>
    <h4>
      Smartphone
    </h4>
    <p
      className="single_request_body"
    >
      Although a downpour briefly interrupted play in the early stages of the second half, Nigeria played with more intent after half-time and went close in the 63rd minute, as John Obi Mikel saw a header saved by Vaclik.
    </p>
    <div
      className="request_footer"
    >
      <a
        className="buttons button_small"
        id="disapprove"
        onClick={[MockFunction]}
      >
        Disapprove
      </a>
      <a
        className="buttons button_small button_right"
        id="approve"
        onClick={[MockFunction]}
      >
        Approve
      </a>
    </div>
  </div>
</div>
`;
